model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,680 @@
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import static org.keycloak.models.UserSessionModel.CORRESPONDING_SESSION_ID;
+import static org.keycloak.models.UserSessionModel.SessionPersistenceState.TRANSIENT;
+import static org.keycloak.models.map.userSession.SessionExpiration.setClientSessionExpiration;
+import static org.keycloak.models.map.userSession.SessionExpiration.setUserSessionExpiration;
+import static org.keycloak.utils.StreamsUtil.paginatedStream;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public class MapUserSessionProvider implements UserSessionProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapUserSessionProvider.class);
+    private final KeycloakSession session;
+    protected final MapKeycloakTransaction<UUID, MapUserSessionEntity, UserSessionModel> userSessionTx;
+    protected final MapKeycloakTransaction<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionTx;
+    private final MapStorage<UUID, MapUserSessionEntity, UserSessionModel> userSessionStore;
+    private final MapStorage<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionStore;
+
+    /**
+     * Storage for transient user sessions which lifespan is limited to one request.
+     */
+    private final Map<UUID, MapUserSessionEntity> transientUserSessions = new HashMap<>();
+
+    public MapUserSessionProvider(KeycloakSession session, MapStorage<UUID, MapUserSessionEntity, UserSessionModel> userSessionStore,
+                                  MapStorage<UUID, MapAuthenticatedClientSessionEntity, AuthenticatedClientSessionModel> clientSessionStore) {
+        this.session = session;
+        this.userSessionStore = userSessionStore;
+        this.clientSessionStore = clientSessionStore;
+        userSessionTx = userSessionStore.createTransaction(session);
+        clientSessionTx = clientSessionStore.createTransaction(session);
+
+        session.getTransactionManager().enlistAfterCompletion(userSessionTx);
+        session.getTransactionManager().enlistAfterCompletion(clientSessionTx);
+    }
+
+    private Function<MapUserSessionEntity, UserSessionModel> userEntityToAdapterFunc(RealmModel realm) {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return (origEntity) -> {
+            if (origEntity.getExpiration() <= Time.currentTime()) {
+                if (Objects.equals(origEntity.getPersistenceState(), TRANSIENT)) {
+                    transientUserSessions.remove(origEntity);
```

Comment:
`transientUserSessions` has `UUID` keys but here `origEntity` is referenced, rendering it a no-op.

Reply:
fixed, thanks for spotting this.